editors/code/src/command/workspace-folder-formatting.ts

```diff
@@ -0,0 +1,228 @@
+import * as cp from "child_process";
+import * as vscode from "vscode";
+
+import { Cmd, Ctx } from "../context";
+import * as output from "../output";
+
+/**
+ * Format a workspace folder
+ *
+ * # Workspace folder selection
+ *
+ * - If 0 workspace folders are open, errors
+ * - If 1 workspace folder is open, automatically uses it
+ * - If >1 workspace folders are open, asks the user to choose
+ *
+ * # Tab closing
+ *
+ * Because we use the embedded Air CLI to perform the formatting, we force all
+ * relevant tabs to be saved and closed before proceeding. This has two main
+ * benefits:
+ *
+ * - The Air CLI can't know about any in-memory changes that may have occurred
+ *   in "dirty" editors. Forcing them to save and close causes any changes
+ *   to be written to disk first, syncing VS Code with the CLI.
+ *
+ * - The LSP server part of Air needs to perfectly track changes for any open
+ *   files. Closing the tabs first switches the server's "source of truth" from
+ *   the client version of the file to the on-disk version of the file, which
+ *   should ensure the LSP server won't have any chance of getting out of sync.
+ *   This may be overkill, since the LSP server seems to handle git branch
+ *   changes well, but it's still not a bad idea.
+ *
+ * # Forward propagation of settings
+ *
+ * If no `air.toml` is tied to a document, then typically in the LSP document
+ * settings are forward propagated from the client to the server, and those are
+ * used during formatting when that document is open. If an `air.toml` does
+ * exist, those settings override document settings.
+ *
+ * Because we go through the Air CLI here, forward propagated document settings
+ * are ignored, so even if you don't have an `air.toml`, you will get Air's
+ * default indentation settings.
+ *
+ * We can't simply require an `air.toml` at the workspace folder root to avoid
+ * confusion here, because you may instead have an `air.toml` at
+ * `{root}/R/air.toml` that would handle any R files in your project. Instead,
+ * we hope this isn't common enough to come up much in practice.
```

Comment:
Notable point of potential confusion, but hopefully not too bad.

Reply:
I was thinking this command should ensure an air.toml file at the project root. Letting the cli potentially pick up config files up the folder hierarchy seems unexpected to me.

Reply:
The comment there explains why it can't just ensure an air.toml is at the project root

Reply:
We currently look _up_ the file hierarchy for `air.toml` but both the CLI and LSP are also going to also look _down_ the file hierarchy for them as well. https://github.com/posit-dev/air/issues/110 When that happens, it will be perfectly reasonable to open, say, the top level `arrow/` workspace to work on the `arrow/r` package nested inside it, and the `air.toml` would be at `arrow/r/air.toml`, i.e. "below" the workspace

Reply:
> Letting the cli potentially pick up config files up the folder hierarchy seems unexpected to me.

Also, both the LSP and the CLI look _up_ the folder hierarchy in the same way. If you open a workspace at `dplyr/R` and your `air.toml` was at `dplyr/`, even though that is "outside" the LSP workspace, we still find and use it. I think that is the only way to have a consistent experience across LSP and CLI.

Reply:
> Running Format Workspace Folder from the arrow/ workspace should be allowed without requiring an air.toml at arrow/air.toml, which wouldn't make sense.

Why not open the `arrow/r` folder and perform the command? That would make more sense to me than do it at top level and perhaps format files in other subfolders using an air.toml file lying around in a parent folder.

Or if we think that's not good UX, just add `Format Workspace Subfolder` and let user quickpick a folder in the project, and that would add the air.toml file at that place.

Reply:
We've decided that as long as we push most people towards running `Air: Setup Workspace Folder` (https://github.com/posit-dev/air/issues/323) first (through documentation), then this will mostly be a non-issue, because an `air.toml` will exist through the setup process.

That allows rare monorepo people like `arrow/` devs to skip that setup step, instead adding configuration to `arrow/r/air.toml`, and still be able to call `Air: Format Workspace Folder` from `arrow/` without being forced into a top level `air.toml`.